parsely/src/main/java/com/parsely/parselyandroid/EngagementManager.java

```diff
@@ -0,0 +1,114 @@
+package com.parsely.parselyandroid;
+
+import java.util.Calendar;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.TimeZone;
+import java.util.Timer;
+import java.util.TimerTask;
+
+/**
+ * Engagement manager for article and video engagement.
+ * <p>
+ * Implemented to handle its own queuing of future executions to accomplish
+ * two things:
+ * <p>
+ * 1. Flushing any engaged time before canceling.
+ * 2. Progressive backoff for long engagements to save data.
+ */
+class EngagementManager {
+
+    private final ParselyTracker parselyTracker;
+    public Map<String, Object> baseEvent;
+    private boolean started;
+    private final Timer parentTimer;
+    private TimerTask waitingTimerTask;
+    private long latestDelayMillis, totalTime;
+    private Calendar startTime;
+    private final UpdateEngagementIntervalCalculator intervalCalculator;
+
+    public EngagementManager(ParselyTracker parselyTracker, Timer parentTimer, long intervalMillis, Map<String, Object> baseEvent, UpdateEngagementIntervalCalculator intervalCalculator) {
+        this.parselyTracker = parselyTracker;
+        this.baseEvent = baseEvent;
+        this.parentTimer = parentTimer;
+        this.intervalCalculator = intervalCalculator;
+        latestDelayMillis = intervalMillis;
+        totalTime = 0;
+        startTime = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
+    }
+
+    public boolean isRunning() {
+        return started;
+    }
+
+    public void start() {
+        scheduleNextExecution(latestDelayMillis);
+        started = true;
+        startTime = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
+    }
+
+    public void stop() {
+        waitingTimerTask.cancel();
+        started = false;
+    }
+
+    public boolean isSameVideo(String url, String urlRef, ParselyVideoMetadata metadata) {
```

Comment:
Suggestion (💡): Consider resolving/suppressing all the warnings related to this method:

1. Suppress: `Unchecked cast: 'java.lang.Object' to 'java.util.Map<java.lang.String,java.lang.Object>'`
2. Resolve: `Method invocation 'equals' may produce 'NullPointerException'`
3. Resolve: `Unboxing of 'baseMetadata.get("duration")' may produce 'NullPointerException'`

FYI: For no.2 there is an easy fix, just flit the `equals(...)`, for example:

**BEFORE**: baseEvent.get("url").equals(url)
**AFTER**: url.equals(baseEvent.get("url")

Reply:
I'd leave this to a different PR - I wanted this PR to be focused on extracting `EngagementManager` and adding minimal set of changes to make it testable.